File: scj-addons/src/main/java/pw/phylame/jem/scj/addons/ScriptRunner.java
package pw.phylame.jem.scj.addons;

import java.io.File;
import java.io.FileReader;

import javax.script.ScriptEngine;

import org.apache.commons.cli.Option;

import lombok.val;
import pw.phylame.qaf.cli.CLIDelegate;
import pw.phylame.qaf.cli.Command;
import pw.phylame.qaf.cli.TypedFetcher;
import pw.phylame.ycl.log.Log;
import pw.phylame.ycl.util.StringUtils;

public class ScriptRunner extends AbstractPlugin {
    private static final String TAG = "CSR";

    private static final String OPTION = "R";

    private static final String DEFAULT_ENGINE = "JavaScript";

    public ScriptRunner() {
        super(new Metadata("ff6369df-2b11-4d9d-80e2-1197fc9e088f", "Script Runner", "1.0", "PW"));
    }

    @Override
    public void init() {
        sci.addOption(Option.builder(OPTION)
                .longOpt("run-script")
                .hasArg()
                .argName(M.tr("runScript.file"))
                .desc(M.tr("help.runScript"))
                .build(), new RunnerCommand());
    }

    private Object detectScriptEngine() {
        try {
            val clazz = Class.forName("javax.script.ScriptEngineManager");
            String name = config.rawFor("csr.engine.name");
            if (StringUtils.isEmpty(name)) {
                name = DEFAULT_ENGINE;
            }
            val engine = clazz.getMethod("getEngineByName", String.class).invoke(clazz.newInstance(), name);
            if (engine == null) {
                app.error(M.tr("runScript.noSuchEngine", name));
            }
            return engine;
        } catch (Exception e) {
            Log.e(TAG, e);
            app.error(M.tr("runScript.unsupported", System.getProperty("java.version")));
            return null;
        }
    }

    private class RunnerCommand extends TypedFetcher<String> implements Command {

        private RunnerCommand() {
            super(OPTION, String.class, null);
        }

        @Override
        public int execute(CLIDelegate delegate) {
            val file = new File(sci.getContext().get(OPTION).toString());
            if (!file.exists()) {
                app.error(app.tr("error.input.notExists", file.getPath()));
                return -1;
            }
            val result = detectScriptEngine();
            if (result == null) {
                return -1;
            }
            try (val reader = new FileReader(file)) {
                val engine = (ScriptEngine) result;
                engine.put("app", app);
                engine.put("sci", sci);
                engine.put("config", config);
                engine.eval(reader);
            } catch (Exception e) {
                app.error(M.tr("runScript.error"), e);
                return -1;
            }
            return 0;
        }
    }

}

File: scj-addons/src/main/kotlin/jem/scj/addon/ScriptRunner.kt
/*
 * Copyright 2017 Peng Wan <phylame@163.com>
 *
 * This file is part of Jem.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jem.scj.addon

import jclp.log.Log
import jem.Book
import jem.epm.EpmManager
import jem.epm.ParserParam
import jem.scj.SCI
import jem.scj.SCIPlugin
import jem.scj.SCISettings
import mala.App
import mala.App.tr
import mala.cli.*
import org.apache.commons.cli.Option
import java.io.File
import javax.script.ScriptEngine
import javax.script.ScriptEngineManager
import javax.script.ScriptException

class ScriptRunner : ListFetcher("R"), SCIPlugin, Command {
//    override val meta = mapOf("name" to "Script Runner")

    override fun init() {
        attachTranslator()
        SCI.newOption("R", "run-script")
                .hasArg()
                .argName(tr("opt.R.arg"))
                .action(this)
        Option.builder()
                .hasArg()
                .longOpt("engine-name")
                .argName(tr("opt.engine.arg"))
                .desc(tr("opt.engineName.desc"))
                .action(StringFetcher("engine-name"))
        Option.builder()
                .hasArg()
                .argName(tr("opt.R.arg"))
                .longOpt("book-filter")
                .desc(tr("opt.bookFilter.desc"))
                .action(StringFetcher("book-filter"))
    }

    @Suppress("UNCHECKED_CAST")
    override fun execute(delegate: CDelegate): Int {
        val paths = SCI["R"] as? List<String> ?: return -1
        var code = 0
        paths.map(::File).forEach {
            code = if (!it.exists()) {
                App.error(tr("err.misc.noFile", it))
                -1
            } else {
                minOf(code, runScript(it))
            }
        }
        return code
    }

    override fun onBookOpened(book: Book, param: ParserParam?) {
        val file = File(SCI["book-filter"]?.toString() ?: return)
        if (!file.exists()) {
            App.error(tr("err.misc.noFile", file))
            return
        }
        runScript(file) {
            put("book", book)
        }
    }

    private fun runScript(file: File, action: (ScriptEngine.() -> Unit)? = null): Int {
        val engine = getScriptEngine(file) ?: return -1
        Log.d(javaClass.simpleName) { "engine $engine detected" }
        action?.invoke(engine)
        try {
            file.reader().use(engine::eval)
        } catch (e: ScriptException) {
            App.error(tr("err.scriptRunner.badScript"), e)
            return -1
        }
        return 0
    }

    private fun getScriptEngine(file: File): ScriptEngine? {
        val engineManager = ScriptEngineManager()
        val name = SCI["engine-name"]?.toString()
        val engine: ScriptEngine?
        if (name != null) {
            engine = engineManager.getEngineByName(name)
            if (engine == null) {
                App.error(tr("err.scriptRunner.noName", name))
                return null
            }
        } else {
            engine = engineManager.getEngineByExtension(file.extension)
            if (engine == null) {
                App.error(tr("err.scriptRunner.noExtension", file.extension))
                return null
            }
        }
        engine.put("app", App)
        engine.put("sci", SCI)
        engine.put("epm", EpmManager)
        engine.put("settings", SCISettings)
        return engine
    }
}
